project/frontend/src/components/ImageSlider.js

```diff
@@ -0,0 +1,223 @@
+ * based on the previous state and a specified action.
+ */
+function reducer(state, action) {
+  switch (action.type) {
+    case Action.SWIPING: {
+      let newLeft = state.lastLeft + action.deltaX;
+      const minLeft = (state.slideCount - 1) * -width;
+      const maxLeft = 0.0;
+      newLeft = Math.min(Math.max(newLeft, minLeft), maxLeft);
+
+      // Update the current slide if user slides
+      // passed multiple slides in one motion.
+      let newSlide = state.slide;
+      const dir = action.deltaX > 0 ? 'right' : 'left';
+      if (dir === 'left') {
+        if (newLeft <= (state.slide + 1) * -width) {
+          newSlide++;
+        } else if (newLeft > state.slide * -width) {
+          newSlide--;
+        }
+      } else {
+        if (newLeft >= (state.slide - 1) * -width) {
+          newSlide--;
+        } else if (newLeft < state.slide * -width) {
+          newSlide++;
+        }
+      }
+
+      return {
+        left: newLeft,
+        lastLeft: state.lastLeft,
+        slide: newSlide,
+        slideCount: state.slideCount,
+      };
+    }
+    case Action.SWIPED_LEFT: {
+      let newSlide = state.slide;
+
+      // Spring slider back to current slide or to next
+      // slide depending on how far the user slid.
+      if (newSlide < state.slideCount - 1) {
+        if (
+          Math.abs(state.left) % width > width / 2 ||
+          action.velocity >= 0.5
+        ) {
+          newSlide++;
+        }
+      }
+
+      const newLeft = newSlide * -width;
+      return {
+        left: newLeft,
+        lastLeft: newLeft,
+        slide: newSlide,
+        slideCount: state.slideCount,
+        shifting: true,
+      };
+    }
+    case Action.SWIPED_RIGHT: {
+      let newSlide = state.slide;
+
+      // Spring slider back to current slide or to previous
+      // slide depending on how far the user slid.
+      if (newSlide > 0) {
+        if (
+          width - (Math.abs(state.left) % width) > width / 2 ||
+          action.velocity >= 0.5
+        ) {
+          newSlide--;
+        }
+      }
+
+      const newLeft = newSlide * -width;
+      return {
+        left: newLeft,
+        lastLeft: newLeft,
+        slide: newSlide,
+        slideCount: state.slideCount,
+        shifting: true,
+      };
+    }
+    case Action.PREV_SLIDE: {
+      const newSlide = state.slide > 0 ? state.slide - 1 : state.slide;
+      const newLeft = newSlide * -width;
+      return {
+        left: newLeft,
+        lastLeft: newLeft,
+        slide: newSlide,
+        slideCount: state.slideCount,
+        shifting: true,
+      };
+    }
+    case Action.NEXT_SLIDE: {
+      const newSlide =
+        state.slide < state.slideCount - 1 ? state.slide + 1 : state.slide;
+      const newLeft = newSlide * -width;
+      return {
+        left: newLeft,
+        lastLeft: newLeft,
+        slide: newSlide,
+        slideCount: state.slideCount,
+        shifting: true,
+      };
+    }
+    default: {
```

Comment:
for the last 4 cases atleast -- could we just pull the return out to the end of the method and just break in the individual cases since they're the same vars ? (might need declaring the vars newLeft etc before the switch)

Reply:
Done! Was able to remove a lot of redundancy by pulling out the return statement!